frame/support/src/migrations.rs

```diff
@@ -15,19 +15,135 @@
 #[cfg(feature = "try-runtime")]
 use sp_std::vec::Vec;
 
+/// Make it easier to write versioned runtime upgrades.
+///
+/// VersionedRuntimeUpgrade allows developers to write migrations without worrying about checking
+/// and setting storage versions. Instead, the developer wraps their migration in this struct which
+/// takes care of version handling using best practices.
+///
+/// It takes 5 type parameters:
+/// - `From`: The version being upgraded from.
+/// - `To`: The version being upgraded to.
+/// - `Inner`: An implementation of `OnRuntimeUpgrade`.
+/// - `Pallet`: The Pallet being upgraded.
+/// - `Weight`: The runtime's RuntimeDbWeight implementation.
+///
+/// When a VersionedRuntimeUpgrades `on_runtime_upgrade`, `pre_upgrade`, or `post_upgrade` method is
+/// called, the on-chain version of the pallet is compared to `From`. If they match, the Inner
+/// equivalent is called and the pallets on-chain version is set to `To` after the migration.
+/// Otherwise, a warning is logged notifying the developer that the upgrade was a noop and should
+/// probably be removed.
+///
+/// Example:
+/// ```ignore
+/// // Migrations to pass to the Executive pallet.
+/// pub type Migrations = (
+/// 	// ...other migrations
+/// 	VersionedRuntimeUpgrade<
+/// 		Const16<4>, // From on-chain version 4
+/// 		Const16<5>, // To on-chain version 5
+/// 		parachains_configuration::migration::v5::MigrateToV5<Runtime>,
+/// 		Configuration,
+/// 		RocksDbWeight,
+/// 	>,
+/// 	VersionedRuntimeUpgrade<
+/// 		Const16<5>, // From on-chain version 5
+/// 		Const16<7>, // To on-chain version 7
+/// 		parachains_configuration::migration::v6::MigrateToV6<Runtime>,
+/// 		Configuration,
+/// 		RocksDbWeight,
+/// 	>,
+/// 	// ...other migrations
+/// );
+/// ```
+pub struct VersionedRuntimeUpgrade<From, To, Inner, Pallet, Weight> {
+	_marker: PhantomData<(From, To, Inner, Pallet, Weight)>,
+}
+
+/// Implementation of the `OnRuntimeUpgrade` trait for `VersionedRuntimeUpgrade`.
+///
+/// Its main function is to perform the runtime upgrade in `on_runtime_upgrade` only if the on-chain
+/// version of the pallets storage matches `From`, and after the upgrade set the on-chian storage to
+/// `To`. If the versions do not match, it writes a log notifying the developer that the migration
+/// is a noop.
+impl<
+		From: Get<u16>,
+		To: Get<u16>,
+		Inner: OnRuntimeUpgrade,
+		Pallet: GetStorageVersion<CurrentStorageVersion = StorageVersion> + PalletInfoAccess,
+		DbWeight: Get<RuntimeDbWeight>,
+	> OnRuntimeUpgrade for VersionedRuntimeUpgrade<From, To, Inner, Pallet, DbWeight>
+{
+	/// Executes pre_upgrade if the migration will run, and passes a boolean to post_upgrade
+	/// indicating if the migration ran.
+	#[cfg(feature = "try-runtime")]
+	fn pre_upgrade() -> Result<Vec<u8>, sp_runtime::TryRuntimeError> {
+		use codec::Encode;
+		let on_chain_version = Pallet::on_chain_storage_version();
+		if on_chain_version == From::get() {
+			Ok((Inner::pre_upgrade(), true).encode())
```

Comment:
`pre_upgrade` must not alter storage. AFAIK we already check that in FRAME, can you double-check please if that is the case?  
Otherwise we would need a `StorageNoopGuard` here. Same for `post_upgrade`.

Reply:
haha I was literally just writing a test trying to lazy and store if pre/post hooks had been called by altering storage and bumped into that `StorageNoopGuard` error, was wondering what that was.. I'll use a different method to test that pre/post hooks are called/not called as expected :)